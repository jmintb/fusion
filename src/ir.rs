use std::{collections::BTreeMap, fmt::Display};

use tracing::debug;

use crate::{
    control_flow_graph::ControlFlowGraph,
    parser::{AccessModes, TSIdentifier, TSValue},
    typed_ast::{
        self, Array, ArrayLookup, Assign, Assignment, Decl, Operation, Return, Type, TypedAst,
        TypedExpression, TypedProgram, While,
    },
};

#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Copy)]
pub struct SSAID(pub usize);
#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Copy)]
pub struct BlockId(pub usize);
#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord)]
pub struct FunctionId(pub TSIdentifier);

#[derive(Clone, Debug)]
pub struct IrProgram {
    pub ssa_variables: BTreeMap<SSAID, Variable>,
    pub blocks: BTreeMap<BlockId, Block>,
    pub access_modes: BTreeMap<SSAID, AccessModes>,
    pub control_flow_graphs: BTreeMap<FunctionId, ControlFlowGraph<BlockId>>,
    pub entry_block: BlockId,
}

impl Display for IrProgram {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str("IR:\n")?;
        for (function_id, control_flow_graph) in self.control_flow_graphs.clone() {
            f.write_fmt(format_args!("fn {}:\n", function_id.0 .0))?;
            for block_id in control_flow_graph.clone().into_ordered_iterator() {
                let block = self.blocks.get(&block_id).unwrap();
                f.write_fmt(format_args!("BLOCK: {}", block_id.0))?;

                for (instruction_count, instruction) in block.instructions.iter().enumerate() {
                    f.write_fmt(format_args!(
                        "{}: {}\n",
                        instruction_count,
                        instruction.to_display_string(&self.ssa_variables)
                    ))?;
                }
            }
        }

        f.write_str("\n")?;

        Ok(())
    }
}

#[derive(Clone, Debug)]
pub struct Block {
    pub instructions: Vec<Instruction>,
}

impl Block {
    fn new() -> Self {
        Self {
            instructions: Vec::new(),
        }
    }
}

#[derive(Clone, Debug)]
pub enum Instruction {
    Assign(SSAID),
    Move(SSAID),
    Borrow(SSAID),
    BorrowEnd(SSAID),
    MutBorrow(SSAID),
    MutBorrowEnd(SSAID),
    Drop(SSAID),
    Call(FunctionId, Vec<(SSAID, AccessModes)>),
}

impl Instruction {
    fn get_inverse_instruction(&self) -> Option<Instruction> {
        match *self {
            Self::Borrow(id) => Some(Self::BorrowEnd(id)),
            Self::MutBorrow(id) => Some(Self::MutBorrowEnd(id)),
            _ => None,
        }
    }

    fn to_display_string(&self, ssa_variables: &BTreeMap<SSAID, Variable>) -> String {
        match self {
            Self::Assign(id) => {
                format!(
                    "= {}_{}",
                    id.0,
                    ssa_variables.get(id).unwrap().original_variable.0
                )
            }
            Self::Move(id) => {
                format!(
                    "move({}_{})",
                    id.0,
                    ssa_variables.get(id).unwrap().original_variable.0
                )
            }

            Self::Borrow(id) => {
                format!(
                    "borrow({}_{})",
                    id.0,
                    ssa_variables.get(id).unwrap().original_variable.0
                )
            }
            Self::BorrowEnd(id) => {
                format!(
                    "borrow_end({}_{})",
                    id.0,
                    ssa_variables.get(id).unwrap().original_variable.0
                )
            }
            Self::MutBorrow(id) => {
                format!(
                    "mut_borrow({}_{})",
                    id.0,
                    ssa_variables.get(id).unwrap().original_variable.0
                )
            }
            Self::MutBorrowEnd(id) => {
                format!(
                    "mut_borrow_end({}_{})",
                    id.0,
                    ssa_variables.get(id).unwrap().original_variable.0
                )
            }
            Self::Drop(id) => {
                format!(
                    "drop({}_{})",
                    id.0,
                    ssa_variables.get(id).unwrap().original_variable.0
                )
            }
            Self::Call(function_id, args) => {
                format!(
                    "@{}({})",
                    function_id.0 .0,
                    args.iter()
                        .map(|(variable_id, access_mode)| format!(
                            "{} {}_{},",
                            access_mode,
                            variable_id.0,
                            ssa_variables.get(variable_id).unwrap().original_variable.0
                        ))
                        .fold(String::new(), |acc, next| format!("{} {}", acc, next))
                )
            }
        }
    }
}

#[derive(Clone, Debug)]
pub struct Variable {
    pub original_variable: TSIdentifier,
    id: SSAID,
}

#[derive(Clone, Debug)]
pub struct IrGenerator {
    ssa_variables: BTreeMap<SSAID, Variable>,
    blocks: BTreeMap<BlockId, Block>,
    id_count: usize,
    entry_block: BlockId,
    access_modes: BTreeMap<SSAID, AccessModes>,
    types: BTreeMap<TSIdentifier, Type>,
    control_flow_graphs: BTreeMap<FunctionId, ControlFlowGraph<BlockId>>,
    current_function: FunctionId,
}

impl IrGenerator {
    pub fn new() -> Self {
        let entry_block = Block {
            instructions: Vec::new(),
        };

        let entry_block_id = BlockId(0);
        let mut blocks = BTreeMap::new();
        blocks.insert(entry_block_id, entry_block);
        // Also starts from main.
        let entry_function = FunctionId(TSIdentifier("main".to_string()));

        Self {
            ssa_variables: BTreeMap::new(),
            blocks,
            id_count: 1,
            entry_block: entry_block_id,
            access_modes: BTreeMap::new(),
            types: BTreeMap::new(),
            control_flow_graphs: BTreeMap::new(),
            current_function: entry_function,
        }
    }

    fn new_ssa_id(&mut self) -> SSAID {
        let new_id = self.id_count;
        self.id_count += 1;
        SSAID(new_id)
    }

    fn add_ssa_variable(&mut self, original_variable_id: TSIdentifier) -> SSAID {
        let id = self.new_ssa_id();
        let ssa_var = Variable {
            original_variable: original_variable_id,
            id,
        };

        self.ssa_variables.insert(id, ssa_var);

        id
    }

    fn new_block_id(&mut self) -> BlockId {
        let new_id = self.id_count;
        self.id_count += 1;
        BlockId(new_id)
    }

    fn add_block(&mut self) -> BlockId {
        let id = self.new_block_id();
        let block = Block::new();
        debug!("add block {}", id.0);

        self.blocks.insert(id, block);

        id
    }

    fn latest_gen_variable(&self, var_id: TSIdentifier) -> Option<SSAID> {
        self.ssa_variables
            .clone()
            .into_values()
            .filter(|ssa_var| ssa_var.original_variable == var_id)
            .fold(None, |acc, ssa_var| {
                if acc.is_none() {
                    Some(ssa_var)
                } else {
                    acc.map(|acc| if acc.id > ssa_var.id { acc } else { ssa_var })
                }
            })
            .map(|ssa_var| ssa_var.id)
    }

    pub fn convert_to_ssa(mut self, program: TypedProgram) -> IrProgram {
        for (key, val) in program.types {
            self.types.insert(key, val);
        }

        let mut current_block = self.entry_block;
        for node in program.ast {
            current_block = self.convert_statement(node, current_block);
        }

        IrProgram {
            ssa_variables: self.ssa_variables,
            blocks: self.blocks,
            entry_block: self.entry_block,
            access_modes: self.access_modes,
            control_flow_graphs: self.control_flow_graphs,
        }
    }

    fn record_cfg_connection(&mut self, parent: BlockId, child: BlockId) {
        debug!("record cfg connection {} -> {}", parent.0, child.0);
        self.control_flow_graphs
            .entry(self.current_function.clone())
            .and_modify(|cfg| cfg.insert_edge(parent, child))
            .or_insert_with(|| {
                let mut cfg = ControlFlowGraph::new(parent);
                cfg.insert_edge(parent, child);
                cfg
            });
    }

    fn convert_statement(&mut self, statement: TypedAst, current_block: BlockId) -> BlockId {
        match statement {
            TypedAst::Assignment(assignment) => self.convert_assignment(assignment, current_block),
            TypedAst::Expression(expression) => {
                self.convert_expression(expression, current_block).0
            }
            TypedAst::Decl(declaration) => self.convert_declaration(declaration, current_block),
        }
    }

    fn convert_assignment(&mut self, assignment: Assignment, current_block: BlockId) -> BlockId {
        let updated_block_id = self.convert_expression(assignment.expression, current_block);
        let ssa_id = self.add_ssa_variable(assignment.id);
        let assign_instruction = Instruction::Assign(ssa_id);
        self.add_instruction(updated_block_id.0, assign_instruction);
        updated_block_id.0
    }

    fn convert_block(&mut self, block: typed_ast::Block, current_block: BlockId) -> BlockId {
        let mut current_block = current_block;
        let mut parent_block = current_block;
        current_block = self.add_block();
        self.record_cfg_connection(parent_block, current_block);

        for statement in block.statements {
            current_block = self.convert_statement(statement, current_block);
            if current_block != parent_block {
                // self.record_cfg_connection(parent_block, current_block);
                parent_block = current_block;
            }
        }

        current_block
    }

    fn get_access_instruction(&self, variable_id: SSAID) -> Instruction {
        match self
            .access_modes
            .get(&variable_id)
            .copied()
            .unwrap_or(AccessModes::Owned)
        {
            AccessModes::Let => Instruction::Borrow(variable_id),
            AccessModes::Owned => Instruction::Move(variable_id),
            AccessModes::Inout => Instruction::Move(variable_id),
        }
    }

    fn convert_expression(
        &mut self,
        expression: TypedExpression,
        current_block: BlockId,
    ) -> (BlockId, Option<SSAID>) {
        let mut current_block = current_block;
        match expression {
            TypedExpression::Call(function_id, arguments) => {
                let Type::Function(_, arg_types, _) =
                    self.types.get(&function_id).cloned().unwrap()
                else {
                    todo!()
                };

                let mut function_args = Vec::new();
                let mut free_instructions = Vec::new();
                let mut setup_instructions = Vec::new();

                for (i, arg) in arguments.into_iter().enumerate() {
                    let (new_current_block, ssa_var) = self.convert_expression(arg, current_block);
                    current_block = new_current_block;
                    if let Some(ssa_var) = ssa_var {
                        self.access_modes.insert(ssa_var, arg_types[i].access_mode);
                        let access_instruction = self.get_access_instruction(ssa_var);
                        setup_instructions.push(access_instruction.clone());
                        self.add_instruction(current_block, access_instruction);
                        function_args.push((ssa_var, arg_types[i].access_mode));

                        if let Some(release_instruction) = self
                            .get_access_instruction(ssa_var)
                            .get_inverse_instruction()
                        {
                            free_instructions.push(release_instruction);
                        }
                    }
                }
                self.add_instruction(
                    current_block,
                    Instruction::Call(FunctionId(function_id), function_args),
                );

                for instruction in free_instructions {
                    self.add_instruction(current_block, instruction);
                }
            }

            TypedExpression::Value(val, _) => match val {
                TSValue::Variable(id) => {
                    let ssa_var = self.latest_gen_variable(id.clone()).unwrap();
                    // self.add_instruction(current_block, self.get_access_instruction(ssa_var));
                    return (current_block, Some(ssa_var));
                }
                _ => (),
            },

            TypedExpression::Struct(_, fields) => {
                for field in fields {
                    (current_block, _) = self.convert_expression(field, current_block);
                }
            }

            TypedExpression::StructFieldRef(var_id, _) => {
                let ssa_var = self.latest_gen_variable(var_id).unwrap();
                self.add_instruction(current_block, self.get_access_instruction(ssa_var));
            }

            TypedExpression::If(if_statement) => {
                (current_block, _) =
                    self.convert_expression(*if_statement.condition, current_block);
                current_block = self.convert_block(if_statement.then_block, current_block);
                if let Some(block) = if_statement.else_block {
                    current_block = self.convert_block(block, current_block);
                }
            }

            TypedExpression::While(While { condition, block }) => {
                let condition_block = self.add_block();
                self.record_cfg_connection(current_block, condition_block);
                current_block = condition_block;
                (current_block, _) = self.convert_expression(*condition, current_block);
                let body_block = self.add_block();
                self.record_cfg_connection(condition_block, body_block);
                let body_block = self.convert_block(block, body_block);
                self.record_cfg_connection(body_block, condition_block);
                let post_loop_block = self.add_block();
                self.record_cfg_connection(condition_block, post_loop_block);
                current_block = post_loop_block;
            }

            TypedExpression::Return(Return { expression }) => {
                if let Some(expression) = expression {
                    (current_block, _) = self.convert_expression(*expression, current_block);
                }
            }

            TypedExpression::Operation(operation) => match *operation {
                Operation::Binary(lhs, _, rhs) => {
                    (current_block, _) = self.convert_expression(lhs, current_block);
                    (current_block, _) = self.convert_expression(rhs, current_block);
                }
            },

            TypedExpression::Assign(Assign { id, expression }) => {
                (current_block, _) = self.convert_expression(*expression, current_block);
                let ssa_id = self.add_ssa_variable(id);
                let assign_instruction = Instruction::Assign(ssa_id);
                self.add_instruction(current_block, assign_instruction);
            }

            TypedExpression::Array(Array { items, .. }) => {
                for item in items {
                    (current_block, _) = self.convert_expression(item, current_block);
                }
            }

            TypedExpression::ArrayLookup(ArrayLookup {
                array_identifier,
                index_expression,
            }) => {
                (current_block, _) = self.convert_expression(*index_expression, current_block);
                let ssa_var = self.latest_gen_variable(array_identifier).unwrap();
                self.add_instruction(current_block, Instruction::Move(ssa_var));
            }
        }

        (current_block, None)
    }

    fn convert_declaration(&mut self, declaration: Decl, current_block: BlockId) -> BlockId {
        match declaration {
            Decl::Function {
                keywords: _,
                id,
                arguments,
                body,
                return_type: _,
            } => {
                if let Some(body) = body {
                    self.current_function = FunctionId(id);
                    let entry_block = self.add_block();

                    for arg in arguments {
                        let ssa_id = self.add_ssa_variable(arg.name);
                        let assign_instruction = Instruction::Assign(ssa_id);
                        self.access_modes.insert(ssa_id, arg.access_mode);
                        self.add_instruction(entry_block, assign_instruction);
                    }

                    let cfg = ControlFlowGraph::new(entry_block);
                    self.control_flow_graphs
                        .insert(self.current_function.clone(), cfg);

                    let _new_block =
                        self.convert_block(typed_ast::Block { statements: body }, entry_block);
                }
            }

            _ => (),
        }

        return current_block;
    }

    fn add_instruction(&mut self, updated_block_id: BlockId, assign_instruction: Instruction) {
        self.blocks
            .get_mut(&updated_block_id)
            .unwrap()
            .instructions
            .push(assign_instruction)
    }
}

#[cfg(test)]
mod test {
    use crate::cli::load_program;

    use super::*;
    use anyhow::Result;
    use rstest::rstest;
    use std::path::PathBuf;

    #[rstest]
    fn test_ir_output(#[files("./ir_test_programs/test_*.ts")] path: PathBuf) -> Result<()> {
        use crate::{parser::parse, typed_ast::type_ast};

        let program = load_program(Some(path.to_str().unwrap().to_string()))?;
        let ast = parse(&program)?;
        let typed_program = type_ast(ast)?;

        let ir_generator = IrGenerator::new();
        let ir_progam = ir_generator.convert_to_ssa(typed_program);

        insta::assert_snapshot!(
            format!(
                "test_well_formed_ir_{}",
                path.file_name().unwrap().to_str().unwrap()
            ),
            format!("{}", ir_progam)
        );
        Ok(())
    }
}
