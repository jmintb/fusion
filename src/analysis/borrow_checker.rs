use anyhow::bail;
use anyhow::Result;
use std::collections::BTreeMap;



use crate::analysis::ir_transformer::IrInterpreter;
use crate::analysis::ir_transformer::TransformContext;
use crate::ir::BlockId;

use crate::ir::{Instruction, IrProgram, SSAID};

#[derive(Debug, Clone)]
pub enum VariableState {
    Ready,
    Borrowed,
    MutBorrowed,
    Moved,
}


#[derive(Clone, Debug)]
pub struct BorrowCheckContext {
    variable_states: BTreeMap<SSAID, VariableState>,
}

impl Default for BorrowCheckContext {
    fn default() -> Self {
        todo!()
    }
}


pub struct BorrowChecker {
    variable_states: BTreeMap<SSAID, VariableState>,
}

impl BorrowChecker {
    pub fn new() -> Self {
        Self {
            variable_states: BTreeMap::new(),
        }
    }

    fn check_instruction(instruction_counter: usize, ctx: &mut TransformContext, block_id: &BlockId,  variable_states: &mut BTreeMap<SSAID, VariableState>) -> Result<usize> {
        
                let block = ctx.scope.blocks.get_mut(block_id).unwrap();
                let Some(instruction) = block.instructions.get(instruction_counter) else {
                    return Ok(0);
                };
                match instruction {
                    Instruction::Call(_function_id, _args ) => {
                        ()
                    }
                    Instruction::Assign(id) => {
                        variable_states.insert(*id, VariableState::Ready);
                    }
                    Instruction::Borrow(id) => match variable_states.get(&id) {
                        Some(VariableState::Ready) => {
                            variable_states.insert(*id, VariableState::Borrowed);
                        }
                        Some(VariableState::Borrowed) => (),
                        Some(VariableState::MutBorrowed) => {
                            bail!(format!(
                                "variable {} in block {} was already mutably borrowed",
                                ctx
                                    .ssa_variables
                                    .get(&id)
                                    .unwrap()
                                    .original_variable
                                    .0 , block_id.0
                            ))
                        }
                        None => bail!(format!("Failed to borrow, Variable {} was not in any state, this should not be possible", ctx
                                    .ssa_variables
                                    .get(&id)
                                    .unwrap()
                                    .original_variable
                                    .0
                            )),

                        Some(VariableState::Moved) => {
                            bail!(format!(
                                "variable {} in block {} was already moved ",
                                ctx
                                    .ssa_variables
                                    .get(&id)
                                    .unwrap()
                                    .original_variable
                                    .0 , block_id.0
                            ))
                        }
                    },
                    Instruction::MutBorrow(id) => match variable_states.get(&id) {
                        Some(VariableState::Ready) => {
                            variable_states.insert(*id, VariableState::MutBorrowed);
                        }
                        e => bail!(format!("can not mut borrow a variable {} which is in state {e:?}",
                            
                            ctx
                                .ssa_variables
                                .get(&id)
                                .unwrap()
                                .original_variable
                                .0
                             ))
                    },
                    Instruction::BorrowEnd(id) => {
                        let old_state =  variable_states.insert(*id, VariableState::Ready );          match old_state {
                            Some(VariableState::Borrowed) => (),
                            e => bail!("can not unborrow a variabled which is in state: {e:?}")
                        }       
                        
                           }

                    Instruction::MutBorrowEnd(id) => {
                        let old_state =  variable_states.insert(*id, VariableState::Ready );          match old_state {
                            Some(VariableState::Ready) => (),
                            e => bail!("can not un mutborrow a variabled which is in state: {e:?}")
                        }       
                        
                           }



                    Instruction::Drop(id) | Instruction::Move(id) => match variable_states.get(&id) {
                        Some(VariableState::Ready) => {
                            variable_states.insert(*id, VariableState::Moved);
                        }
                        e => bail!(format!(
                            "variable {} in block {} was already {e:?}",
                            ctx
                                .ssa_variables
                                .get(&id)
                                .unwrap()
                                .original_variable
                                .0
                    , block_id.0
                        )),
                    },
                }
                
                Ok(instruction_counter + 1)
    }


    pub fn check(&mut self, ir_program: &IrProgram) -> Result<()> {
        for function_id in ir_program.control_flow_graphs.keys() {
        let interpreter = IrInterpreter::<BTreeMap<SSAID, VariableState>>::new(ir_program.control_flow_graphs.get(function_id).cloned().unwrap(), ir_program.clone());
        interpreter.transform(&mut Self::check_instruction)?;
        }

        Ok(())
    }


    }

#[cfg(test)]
mod test {
    use crate::cli::load_program;

    use super::*;
    use crate::{ir::IrGenerator, parser::parse, typed_ast::type_ast};
    use anyhow::Result;
    use rstest::rstest;
    use std::path::PathBuf;
    use tracing::debug;

    #[rstest]
    #[test_log::test]
    fn test_borrow_checker(#[files("./ir_test_programs/test_*.ts")] path: PathBuf) -> Result<()> {
        let program = load_program(Some(path.to_str().unwrap().to_string()))?;
        let ast = parse(&program)?;
        let typed_program = type_ast(ast)?;

        let ir_generator = IrGenerator::new();
        let ir_program = ir_generator.convert_to_ssa(typed_program);
        let liveness = crate::analysis::liveness_analysis::calculate_livenss(&ir_program)?;
        let ir_program = crate::analysis::free_dead_resources::insert_free(liveness, ir_program);
        debug!("transformed IR program: {}", ir_program);
        debug!("CFG: {:#?}", ir_program.control_flow_graphs);
        let mut borrow_checker = BorrowChecker::new();
        let analysis_result = borrow_checker.check(&ir_program);

        insta::assert_snapshot!(
            format!(
                "test_borrow_checking{}",
                path.file_name().unwrap().to_str().unwrap()
            ),
            format!("{:#?}", analysis_result)
        );
        Ok(())
    }
}
