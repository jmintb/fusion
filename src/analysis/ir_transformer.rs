use std::collections::{BTreeMap, BTreeSet, VecDeque};

use crate::ir::{Block, IrProgram, Variable, SSAID};
use crate::ast::nodes::AccessModes;

use crate::{control_flow_graph::ControlFlowGraph, ir::BlockId};
use anyhow::Result;
use tracing::debug;

use super::borrow_checker::VariableState;

#[derive(Clone, Debug)]
pub struct IrScope {
    pub blocks: BTreeMap<BlockId, Block>,
    pub control_flow_graph: ControlFlowGraph<BlockId>,
}

#[derive(Clone, Debug)]
pub struct IrInterpreter<Ctx> {
    scope: IrScope,
    control_flow_graph: ControlFlowGraph<BlockId>,
    block_states: BTreeMap<BlockId, Ctx>,
    pub ssa_variables: BTreeMap<SSAID, Variable>,
    pub access_modes: BTreeMap<SSAID, AccessModes>,
    context: Ctx,
    reverse_traversel: bool,
}

pub struct ReverseCycleAwareBlockIterator {
    qeueue: Vec<Vec<BlockId>>,
}

impl ReverseCycleAwareBlockIterator {
    fn new(control_flow_graph: ControlFlowGraph<BlockId>) -> Self {
        Self {
            qeueue: {
                let mut successors = control_flow_graph
                    .cycle_aware_successors(&control_flow_graph.entry_point)
                    .unwrap();
                successors.insert(0, vec![control_flow_graph.entry_point]);
                successors
            },
        }
    }
}

impl Iterator for ReverseCycleAwareBlockIterator {
    type Item = Vec<BlockId>;

    fn next(&mut self) -> Option<Self::Item> {
        self.qeueue.pop()
    }
}

pub struct IrBlockIterator {
    queue: VecDeque<BlockId>,
    visited_blocks: BTreeSet<BlockId>,
    control_flow_graph: ControlFlowGraph<BlockId>,
    visit_again: BTreeSet<BlockId>,
}

impl IrBlockIterator {
    fn new(control_flow_graph: ControlFlowGraph<BlockId>) -> Self {
        let mut queue = VecDeque::new();
        queue.push_back(control_flow_graph.entry_point);
        Self {
            control_flow_graph,
            queue,
            visited_blocks: BTreeSet::new(),
            visit_again: BTreeSet::new(),
        }
    }
}

impl Iterator for IrBlockIterator {
    type Item = BlockId;

    fn next(&mut self) -> Option<Self::Item> {
        let mut next = self.queue.pop_front()?;
        while self.visited_blocks.contains(&next) && !self.visit_again.contains(&next) {
            next = self.queue.pop_front()?;
        }

        // self.visited_blocks.remove(&next);

        debug!("next: {}, queue {:?}", next.0, self.queue);

        // TODO: cycles  seem to be handled , but we need a rigorous test and to avoid the extra visits currently happening.
        if !self.visited_blocks.contains(&next) {
            for predecessor in self.control_flow_graph.direct_predecessors(&next.clone()) {
                if !self.visited_blocks.contains(&predecessor) {
                    if self.control_flow_graph.dominates(next, predecessor) {
                        let predecessors_preds = self
                            .control_flow_graph
                            .predecessors(&predecessor)
                            .unwrap()
                            .into_iter()
                            .flatten()
                            .collect::<Vec<Self::Item>>();

                        debug!(
                            "pred preds: {:?} \n  {} {}",
                            predecessors_preds, predecessor, next
                        );

                        let Some(loop_start_pos) =
                            predecessors_preds.iter().position(|pred| *pred == next)
                        else {
                            break;
                        };

                        let mut loop_blocks = predecessors_preds;
                        loop_blocks.split_off(loop_start_pos);
                        loop_blocks.reverse();
                        loop_blocks.push(predecessor);

                        debug!("cycle blocks: {:?}", loop_blocks);
                        assert!(!loop_blocks.contains(&next));

                        let mut loop_successors: Vec<BlockId> = self
                            .control_flow_graph
                            .successors(&loop_blocks[0])
                            .unwrap()
                            .into_iter()
                            .flatten()
                            .collect();

                        let cycle_completetion_position = loop_successors
                            .iter()
                            .position(|pred| *pred == next)
                            .unwrap();

                        loop_successors.split_off(cycle_completetion_position);

                        // loop_successors
                        //     .clone()
                        //     .into_iter()
                        //     .for_each(|block| loop_blocks.push(block));

                        self.visited_blocks.insert(next);
                        self.visit_again.insert(next);
                        if loop_blocks.len() == 1 {
                            self.queue.push_front(loop_blocks[0]);
                            self.queue.push_front(next);
                            self.visit_again.insert(loop_blocks[0]);
                            next = loop_blocks[0];
                            continue;
                        }

                        let (loop_start, loop_blocks) = loop_blocks.split_first().unwrap();
                        debug!(
                            "loop successor blocks: {:?} {:?} {} {}",
                            loop_successors, loop_blocks, loop_start, predecessor
                        );
                        let mut loop_blocks = loop_blocks.to_vec().clone();

                        loop_blocks.reverse();

                        self.queue.push_front(*loop_start);
                        for loop_predecessor in loop_blocks.clone() {
                            self.queue.push_front(loop_predecessor);
                            self.visit_again.insert(loop_predecessor);
                        }

                        self.queue.push_front(next);
                        self.queue.push_front(*loop_start);
                        self.visit_again.insert(*loop_start);
                        for loop_predecessor in loop_blocks {
                            self.queue.push_front(loop_predecessor);
                        }
                        debug!("skipping {next} {loop_start}");

                        // TODO: remove predecssor from grandchildren, as we end up visiting it again.
                        // This logic tricky, as we need to ensure we revisit the dominated block after visting the dominator.
                        next = *loop_start;
                        break; // TODO: Continuing probably doesn't make sense here as we could have multiple unvisited predecessors I think.
                    } else {
                        // todo!(
                        //     "this should not be possible, but how do we handle it? ,{:#?} {} {}",
                        //     self.control_flow_graph,
                        //     predecessor.0,
                        //     next.0
                        // );
                    }
                }
            }
        }
        debug!("processing block: {}", next.0);

        for &grand_child in self
            .control_flow_graph
            .graph
            .get(&next)
            .unwrap_or(&Vec::new())
        {
            if !self.visited_blocks.contains(&grand_child) {
                self.queue.push_back(grand_child);
            }
        }

        self.visited_blocks.insert(next);

        // sleep(Duration::from_secs(1));

        Some(next)
    }
}

impl IntoIterator for ControlFlowGraph<BlockId> {
    type Item = BlockId;

    type IntoIter = IrBlockIterator;

    fn into_iter(self) -> Self::IntoIter {
        IrBlockIterator::new(self)
    }
}

impl ControlFlowGraph<BlockId> {
    fn into_reverse_cycle_aware_iterator(self) -> ReverseCycleAwareBlockIterator {
        ReverseCycleAwareBlockIterator::new(self)
    }
}

pub struct TransformContext {
    pub scope: IrScope,
    pub ssa_variables: BTreeMap<SSAID, Variable>,
    pub access_modes: BTreeMap<SSAID, AccessModes>,
    pub variable_states: BTreeMap<SSAID, VariableState>,
}

type TransformFn<Ctx> =
    dyn FnMut(usize, &mut TransformContext, &BlockId, &mut Ctx) -> Result<usize>;

impl<Ctx: Clone + Default> IrInterpreter<Ctx> {
    pub fn new(control_flow_graph: ControlFlowGraph<BlockId>, program: IrProgram) -> Self {
        let scope = IrScope {
            blocks: program.blocks,
            control_flow_graph: control_flow_graph.clone(),
        };
        Self {
            control_flow_graph,
            scope,
            ssa_variables: program.ssa_variables,
            access_modes: program.access_modes,
            block_states: BTreeMap::new(),
            context: Ctx::default(),
            reverse_traversel: false,
        }
    }

    pub fn new_reversed(control_flow_graph: ControlFlowGraph<BlockId>, program: IrProgram) -> Self {
        let scope = IrScope {
            blocks: program.blocks,
            control_flow_graph: control_flow_graph.clone(),
        };
        Self {
            control_flow_graph,
            scope,
            ssa_variables: program.ssa_variables,
            access_modes: program.access_modes,
            block_states: BTreeMap::new(),
            context: Ctx::default(),
            reverse_traversel: true,
        }
    }

    pub fn transform(mut self, transform_fn: &mut TransformFn<Ctx>) -> Result<Ctx> {
        let mut ctx = TransformContext {
            scope: self.scope.clone(),
            ssa_variables: self.ssa_variables.clone(),
            access_modes: self.access_modes.clone(),
            variable_states: BTreeMap::new(),
        };

        if self.reverse_traversel {
            for block in self
                .control_flow_graph
                .clone()
                .into_reverse_cycle_aware_iterator()
                .flatten()
            {
                self.transform_block(block, transform_fn, &mut ctx)?;
            }
        } else {
            // for block in self.control_flow_graph.clone() {
            //     debug!("block order {}", block);
            // }
            for block in self.control_flow_graph.clone() {
                self.transform_block(block, transform_fn, &mut ctx)?;
            }
        }

        Ok(self.context)
    }

    fn transform_block(
        &mut self,
        block_id: BlockId,
        transform_fn: &mut TransformFn<Ctx>,
        ctx: &mut TransformContext,
    ) -> Result<()> {
        let block = self.scope.blocks.get_mut(&block_id).unwrap();
        if block.instructions.is_empty() && self.reverse_traversel {
            return Ok(());
        }

        let mut instruction_counter = if self.reverse_traversel {
            block.instructions.len() - 1
        } else {
            0
        };

        loop {
            instruction_counter =
                transform_fn(instruction_counter, ctx, &block_id, &mut self.context)?;

            if (instruction_counter == 0 && !self.reverse_traversel)
                || (self.reverse_traversel && instruction_counter == block.instructions.len())
            {
                break;
            }
        }

        Ok(())
    }
}
