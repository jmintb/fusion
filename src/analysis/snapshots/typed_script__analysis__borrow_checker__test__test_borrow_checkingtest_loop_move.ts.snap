---
source: src/analysis/borrow_checker.rs
expression: "format!(\"{:#?}\", analysis_result)"
---
Err(
    "variable value was already Some(Moved)",
)
